### PR TITLE
Obasilakis/endpoint for individual voter rewards

### PR DESCRIPTION
**Pull request by obasilakis**
### Description

First added a rewards module to the api with 2 actions: `getvoterrewardsforgroup` and `getvoterrewards`. The latter will probably be more popular, but I added the former in case we want to create a page with the rewards for a specific group.

To calculate voter rewards for a group, the voter's share for the group has to be calculated for 1 block before the epoch block. To that end I added `Indexer.Fetcher.CeloValidatorGroupVotes`, an indexer that queries the `fetch_epoch_rewards` table (populated by `Explorer.Repo.Migrations.AddInitialCeloPendingEpochOperationsGroupVotes`) and saves the total votes for a group 1 block before the epoch block to the `celo_validator_group_votes` table. One thing to note here is that the above-mentioned migration is almost identical to `Explorer.Repo.Migrations.AddInitialCeloPendingEpochOperations`. I am not sure whether an old migration can be changed without causing issues, that's why I added a new one.

`Explorer.Celo.VoterRewards.calculate(voter_address_hash, from_date, to_date)` is responsible for calculating the rewards for all groups. It gets all the groups for which the voter has activated CELO, calls `Explorer.Celo.VoterRewardsForGroup.calculate(voter_address_hash, group_address_hash)` on each of the groups, filters by date and returns the rewards in a useful structure. `voter_rewards_for_group` became part of the explorer configuration and returns a different module depending on the environment (test will get `Explorer.Fakes.VoterRewardsForGroup` to avoid setting up the db again for `Explorer.Celo.VoterRewardsForGroup.calculate(voter_address_hash, group_address_hash)`'s test).

### Tests and Docs

Tests cover new functionality and documentation was added for the 2 new endpoints.

### CODE REVIEW

File: apps/explorer/test/explorer/celo/voter_rewards_for_group_test.exs [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    test "sums a voter's activated and revoked CELO for the previous day of the block passed" do
+      voter_activated_or_revoked = [
+        %{
+          block_hash: %Hash{
+            byte_count: 32,
+            bytes: <<0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0>>
+          },
+          block_number: 10_692_863,
+          amount_activated_or_revoked: 650,
+          event: "0x45aac85f38083b18efe2d441a65b9c1ae177c78307cb5a5d4aec8f7dbcaeabfe",
```
Comment (rkachowski) [inline apps/explorer/test/explorer/celo/voter_rewards_for_group_test.exs:106]: why are we using topics here?

File: apps/indexer/lib/indexer/fetcher/celo_validator_group_votes.ex [hunk 2]
```diff
@@ -0,0 +1,121 @@
+defmodule Indexer.Fetcher.CeloValidatorGroupVotes do
```
Comment (rkachowski) [inline apps/indexer/lib/indexer/fetcher/celo_validator_group_votes.ex:1]: I feel we have too many epoch related fetcher modules and we will need to refactor these into a single CeloEpochData fetcher at some point

File: apps/explorer/lib/explorer/chain.ex [hunk 3]
```diff
@@ -7874,13 +7926,46 @@ defmodule Explorer.Chain do
...
+      |> Enum.each(fn reward ->
+        Chain.falsify_or_delete_celo_pending_epoch_operation(reward.block_hash, :fetch_epoch_rewards)
+      end)
+
+      {:ok, success}
+    end)
+    |> Explorer.Repo.transaction()
+  end
+
+  def import_group_votes_and_delete_pending_celo_epoch_operations(import_params, success) do
```
Comment (rkachowski) [inline apps/explorer/lib/explorer/chain.ex:7938]: Let's stop using Ecto.Multi in the Chain context and use the import runner functionality instead. Database operations are not related to Blockchain functionality and adding features here is adding to the mess that is this file.

File: apps/explorer/lib/explorer/chain.ex [hunk 4]
```diff
@@ -7860,10 +7894,28 @@ defmodule Explorer.Chain do
     |> Repo.one()
   end
 
-  @spec delete_celo_pending_epoch_operation(Hash.Full.t()) :: CeloPendingEpochOperation.t()
-  def delete_celo_pending_epoch_operation(block_hash) do
+  @spec falsify_or_delete_celo_pending_epoch_operation(
+          Hash.Full.t(),
+          :fetch_epoch_rewards | :fetch_validator_group_data
+        ) :: CeloPendingEpochOperation.t()
+  def falsify_or_delete_celo_pending_epoch_operation(block_hash, operation_type) do
```
Comment (rkachowski) [inline apps/explorer/lib/explorer/chain.ex:7901]: this isn't relevant to the chain. we can do this inside of the CeloPendingEpochOperation import runner.

Similar to how it is done with internal transactions https://github.com/celo-org/blockscout/blob/a801b8f2f5d8d5c4da27c37ceeb3bf26397d36f7/apps/explorer/lib/explorer/chain/import/runner/internal_transactions.ex#L39

File: apps/indexer/lib/indexer/fetcher/celo_epoch_rewards.ex [hunk 5]
```diff
@@ -54,7 +54,7 @@ defmodule Indexer.Fetcher.CeloEpochRewards do
   def fetch_from_blockchain(blocks) do
     blocks
     |> Enum.map(fn block ->
-      case AccountReader.validator_group_reward_data(block) do
+      case AccountReader.epoch_reward_data(block) do
```
Comment (rkachowski) [inline apps/indexer/lib/indexer/fetcher/celo_epoch_rewards.ex:57]: is this fetching all epoch reward data or just voter rewards?

File: apps/indexer/lib/indexer/fetcher/celo_validator_group_votes.ex [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+  end
+
+  def import_items(rewards) do
+    {failed, success} =
+      Enum.reduce(rewards, {[], []}, fn
+        %{error: _error} = reward, {failed, success} ->
+          {[reward | failed], success}
+
+        reward, {failed, success} ->
+          changeset = CeloValidatorGroupVotes.changeset(%CeloValidatorGroupVotes{}, reward)
```
Comment (rkachowski) [inline apps/indexer/lib/indexer/fetcher/celo_validator_group_votes.ex:85]: The indexer should not be handling changesets

File: apps/indexer/test/indexer/fetcher/celo_validator_group_votes_test.exs [hunk 7]
```diff
@@ -0,0 +1,211 @@
+defmodule Indexer.Fetcher.CeloValidatorGroupVotesTest do
+  # MUST be `async: false` so that {:shared, pid} is set for connection to allow CoinBalanceFetcher's self-send to have
```
Comment (rkachowski) [inline apps/indexer/test/indexer/fetcher/celo_validator_group_votes_test.exs:2]: This file doesn't reference or utilise the CoinBalanceFetcher

**Review by rkachowski — CHANGES_REQUESTED** [review]

File: apps/explorer/lib/explorer/celo/voter_rewards.ex [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+  alias Election.{
+    EpochRewardsDistributedToVotersEvent,
+    ValidatorGroupActiveVoteRevokedEvent,
+    ValidatorGroupVoteActivatedEvent
+  }
+
+  def calculate(voter_address_hash, from_date, to_date) do
+    from_date =
+      case from_date do
+        nil -> ~U[2020-04-22 16:00:00.000000Z]
```
[the remark this answers is not in the document]
Comment (obasilakis) [reply apps/explorer/lib/explorer/celo/voter_rewards.ex:26]: yeah

File: apps/indexer/lib/indexer/fetcher/celo_epoch_rewards.ex (hunk 5 above)
Comment (obasilakis) [reply apps/indexer/lib/indexer/fetcher/celo_epoch_rewards.ex:57]: all epoch rewards. it was a mistake from a previous PR

**Review by rkachowski — APPROVED** [review]

**Pull request merged**